src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallation.m

```diff
@@ -73,11 +67,50 @@ - (NSData *) toJsonData {
     NSDictionary * dictionary = @{
        @"installationId" : self.installationID,
        @"platform" : self.platform,
-       @"pushChannel" : self.pushChannel
+       @"pushChannel" : self.pushChannel,
+       @"tags" : self.tags ?: @""
     };
     
     return [NSJSONSerialization dataWithJSONObject:dictionary
     options:NSJSONWritingPrettyPrinted error:nil];
 }
 
+- (BOOL) addTags:(NSArray<NSString *> *)tags {
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"^[a-zA-Z0-9_@#\\.:\\-]{1,120}$"
+    options:NSRegularExpressionCaseInsensitive error:nil];
+    NSMutableArray *tmpTags = [NSMutableArray arrayWithArray:self.tags];
+    
+    for(NSString *tag in tags) {
+        if(![tmpTags containsObject:tag]) {
+            if([regex numberOfMatchesInString:tag options:0 range:NSMakeRange(0, tag.length)] > 0)
+            {
+                [tmpTags addObject:tag];
+            } else {
+                NSLog(@"Invalid tag: %@", tag);
```

Comment:
Should we always return `YES` even if the tag is invalid?

Reply:
We are validating all the input tags in a loop, so, valid will be added, invalid will be skipped. Seems we should not return NO if some tag does not match criteria, should we?

Reply:
Otherwise we should abort `addTags` if at least one tag is incorrect and return NO. 
@mpodwysocki thoughts?

Reply:
Yeah, I think we should abort the whole operation if they are invalid

Reply:
Done